src/main/scala/chiselverify/approximation/CompoundMetrics.scala

```diff
@@ -0,0 +1,520 @@
+    */
+  final case class SSIM(maxVal: Option[Double] = None, pixelWidth: Int = 8)
+    extends ImageMetric(maxVal, pixelWidth) {
+    /** 
+      * Stability constants
+      */
+    private val k1: Double = 0.01
+    private val k2: Double = 0.03
+    private val c1: Double = scala.math.pow(k1 * Max.toDouble, 2.0)
+    private val c2: Double = scala.math.pow(k2 * Max.toDouble, 2.0)
+
+    /** 
+      * Generate a Gaussian window of size `n` by `n`
+      * 
+      * @param n side length of the window
+      * @param std the standard deviation of the Gaussian distribution
+      * @return a 2D matrix with a symmetric Gaussian `n` by `n` window
+      * 
+      * Follows the definition used in Wang et al [2004]
+      * Image Quality Assessment: From Error Visibility to Structural Similarity
+      */
+    private def gaussian(n: Int, std: Double = 1.5): Iterable[Iterable[Double]] = {
+      require(n > 0)
+      val denum  = 2 * scala.math.pow(std, 2.0)
+      val coords = (0 until n).map(i => i - ((n-1)/2) - (if ((n & 0x1) == 0) 0.5 else 0.0))
+      val gauss  = coords.map { x => coords.map { y =>
+        scala.math.exp(-((scala.math.pow(x, 2.0) + scala.math.pow(y, 2.0)) / denum))
+      }}
+      val sum = gauss.foldLeft(0.0) { case (acc, row) => acc + row.sum }
+      gauss.map(row => row.map(_ / sum))
+    }
+
+    /** 
+      * Compute the element-wise product of two identically-dimensioned images
+      * 
+      * @param vs1 first image
+      * @param vs2 second image
+      * @return the element-wise product of `vs1` and `vs2`
+      */
+    private def mm(vs1: Iterable[Iterable[Double]],
+                   vs2: Iterable[Iterable[Double]]): Iterable[Iterable[Double]] = {
+      require(vs1.size == vs2.size && vs1.zip(vs2).forall { case (row1, row2) => row1.size == row2.size },
+        "the matrices must have identical dimensions")
+      vs1.zip(vs2).map { case (row1, row2) => row1.zip(row2).map { case (pxl1, pxl2) => pxl1 * pxl2 }}
+    }
+
+    /** 
+      * Compute the cross-correlation of two identically-dimensioned images
+      * 
+      * @param vs1 first image
+      * @param vs2 second image
+      * @return the cross-correlation of `vs1` and `vs2`
+      * 
+      * Assumes the implementation defined at
+      * https://observablehq.com/@lemonnish/cross-correlation-of-2-matrices
+      */
+    private def xcorr(vs1: Iterable[Iterable[Double]],
+                      vs2: Iterable[Iterable[Double]]): Double = mm(vs1, vs2).flatten.sum
+
+    /** 
+      * Generates windows of size `n` by `n` of an image
+      * 
+      * @param vs image to generate windows from
+      * @param n side length of the windows
+      * @return a list of windows of size `n` by `n` sliced from `vs`
+      * 
+      * @note Assumes the image is at least `n` by `n` pixels
+      */
+    private def windows(vs: Iterable[Iterable[Double]], n: Int):
+      Iterable[Iterable[Iterable[Double]]] = {
+      assume(vs.size >= n && vs.forall(_.size >= n))
```

Comment:
Having assumptions like this in the code is quite risky, are they really necessary? I think that in most cases your assumptions should be assertions to avoid any false positives when checking things.

Reply:
Changed to an assertion.